File: binary-tree-maximum-path-sum.cc
// Binary Tree Maximum Path Sum
class Solution {
public:
  int maxPathSum(TreeNode *root) {
    int ans = INT_MIN, mx = INT_MIN;
    f(root, ans, mx);
    return mx < 0 ? mx : ans;
  }
  int f(TreeNode *p, int &ans, int &mx) {
    if (! p) return 0;
    mx = max(mx, p->val);
    int l = f(p->left, ans, mx), r = f(p->right, ans, mx);
    ans = max(ans, l+p->val+r);
    return max(max(l, r)+p->val, 0);
  }
};

File: binary-tree-maximum-path-sum.nim
type
  TreeNode {.importc: "struct TreeNode".} = object
    val: cint
    left, right: ptr TreeNode

proc maxPathSum(root: ptr TreeNode): int {.exportc.} =
  var res: int = root.val
  proc dfs(x: ptr TreeNode): int =
    if x == nil: return 0
    let l = max(dfs(x.left), 0)
    let r = max(dfs(x.right), 0)
    res = max(res, l+x.val+r)
    max(l, r)+x.val

  discard dfs(root)
  res
